### PR TITLE
Wrap `fromJson` with logging

### PR DESCRIPTION
**Pull request by dickermoshe**

### CODE REVIEW

**Comment by dickermoshe** [comment]
@walsha2 
 Take a look, very simple PR.
 Going to work on the raw strings now

File: lib/src/utils/json_logging.dart [hunk 1]
```diff
@@ -0,0 +1,22 @@
+/// Wraps a function that parses a JSON map to a specific type and logs any errors that occur during parsing.
+T fromJsonWithLogging<T, B>(B json, T Function(B) getJson) {
+  try {
+    return getJson(json);
+  } catch (e) {
+    if (e is _ParsingJsonException) {
+      rethrow;
+    } else {
+      // ignore: avoid_print
+      throw _ParsingJsonException("Failed to parse $T: $json\n$e");
```
Comment (walsha2) [inline lib/src/utils/json_logging.dart:10]: Do you have a sample before/after exception to see the improved reporting?

Comment (dickermoshe) [reply lib/src/utils/json_logging.dart:10]: Will remove...

Comment (dickermoshe) [reply lib/src/utils/json_logging.dart:10]: <details><summary>Original</summary>
<p>

```
Building package executable... 
Built openapi_spec:openapi_spec.
Path to the OpenAPI schema file: .\openai.yaml
Unhandled exception:
type 'int' is not a subtype of type 'String' in type cast #  Where ?????
#0      _$$TagImplFromJson (package:openapi_spec/src/open_api/index.g.dart:1463:26)
#1      new _$TagImpl.fromJson (package:openapi_spec/src/open_api/index.freezed.dart:13721:7)
#2      _$TagFromJson (package:openapi_spec/src/open_api/index.freezed.dart:13583:15)
#3      new Tag.fromJson (package:openapi_spec/src/open_api/tag.dart:25:54)
#4      new OpenApi.fromJson.<anonymous closure> (package:openapi_spec/src/open_api/spec.dart:213:59)
#5      MappedListIterable.elementAt (dart:_internal/iterable.dart:425:31)
#6      ListIterator.moveNext (dart:_internal/iterable.dart:354:26)
#7      new _GrowableList._ofEfficientLengthIterable (dart:core-patch/growable_array.dart:189:27)
#8      new _GrowableList.of (dart:core-patch/growable_array.dart:150:28)
#9      new List.of (dart:core-patch/array_patch.dart:39:18)
#10     ListIterable.toList (dart:_internal/iterable.dart:224:7)
#11     new OpenApi.fromJson (package:openapi_spec/src/open_api/spec.dart:213:72)
#12     new OpenApi.fromString (package:openapi_spec/src/open_api/spec.dart:164:20)
#13     GenerateCommand.run (package:openapi_spec/src/commands/generate.dart:119:24)
#14     CommandRunner.runCommand (package:args/command_runner.dart:212:27)
#15     CompletionCommandRunner.runCommand (package:cli_completion/src/command_runner/completion_command_runner.dart:89:18)
#16     OpenapiSpecGeneratorCommandRunner.runCommand (package:openapi_spec/src/command_runner.dart:116:30)
#17     OpenapiSpecGeneratorCommandRunner.run (package:openapi_spec/src/command_runner.dart:60:20)
#18     main (file:///C:/Users/dicke/DickerSystems/openapi_spec/bin/openapi_spec.dart:6:66)
#19     _delayEntrypointInvocation.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:295:33)
#20     _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:184:12)
```

</p>
</details> 


<details><summary>Without custom exception</summary>
<p>

When child `fromJson` rethrows the parse error, the parent `fromJson` also logs it.

```
Building package executable... 
Built openapi_spec:openapi_spec.
Path to the OpenAPI schema file: .\a.yaml
Failed to parse Tag: {name: 456, description: Learn how to turn audio into text.}
type 'int' is not a subtype of type 'String' in type cast
Failed to parse OpenApi: {openapi: 3.0.0, info: {title: OpenAI API, description: The OpenAI REST API. Please see https://platform.openai.com/docs/api-reference for more details., version: 2.0.0, termsOfService: https://openai.com/policies/terms-of-use, contact: {name: OpenAI Support, url: https://help.openai.com/}, license: {name: MIT, url: https://github.com/openai/openai-openapi/blob/master/LICENSE}}, servers: [{url: https://api.openai.com/v1}], tags: [{name: 456, description: Learn how to turn audio into text.}, {name: Chat, description: Given a list of messages comprising a conversation, the model will return a response.}, {name: Completions, description: Given a prompt, the model will return one or more predicted completions, and can also return the probabilities of alternative tokens at each position.}, {name: Embeddings, description: Get a vector representation of a given input that can be easily consumed by machine learning models and algorithms.}, {name: Fine-tuning, description: Manage fine-tuning jobs to tailor a model to your specific training data.}, {name: Files, description: Files are used to upload documents that can be used with features like fine-tuning.}, {name: Images, 
...........
We recommend transitioning to the updating [fine-tuning API](/docs/guides/fine-tuning)
, sections: [{type: object, key: FineTune, path: object}, {type: endpoint, key: createFineTune, path: create}, {type: endpoint, key: listFineTunes, path: list}, {type: endpoint, key: retrieveFineTune, path: retrieve}, {type: endpoint, key: cancelFineTune, path: cancel}, {type: object, key: FineTuneEvent, path: event-object}, {type: endpoint, key: listFineTuneEvents, path: list-events}]}, {id: edits, title: Edits, deprecated: true, description: Given a prompt and an instruction, the model will return an edited version of the prompt.
, sections: [{type: object, key: CreateEditResponse, path: object}, {type: endpoint, key: createEdit, path: create}]}]}}
type 'int' is not a subtype of type 'String' in type cast
Unhandled exception:
type 'int' is not a subtype of type 'String' in type cast
#0      _$$TagImplFromJson (package:openapi_spec/src/open_api/index.g.dart:1463:26)
#1      new _$TagImpl.fromJson (package:openapi_spec/src/open_api/index.freezed.dart:13721:7)
#2      _$TagFromJson (package:openapi_spec/src/open_api/index.freezed.dart:13583:15)
#3      fromJsonWithLogging (package:openapi_spec/src/utils/json_logging.dart:4:19)
#4      new Tag.fromJson (package:openapi_spec/src/open_api/tag.dart:26:7)
#5      new OpenApi.fromJson.<anonymous closure>.<anonymous closure> (package:openapi_spec/src/open_api/spec.dart:215:59)
#6      MappedListIterable.elementAt (dart:_internal/iterable.dart:425:31)
#7      ListIterator.moveNext (dart:_internal/iterable.dart:354:26)
#8      new _GrowableList._ofEfficientLengthIterable (dart:core-patch/growable_array.dart:189:27)
#9      new _GrowableList.of (dart:core-patch/growable_array.dart:150:28)
#10     new List.of (dart:core-patch/array_patch.dart:39:18)
#11     ListIterable.toList (dart:_internal/iterable.dart:224:7)
#12     new OpenApi.fromJson.<anonymous closure> (package:openapi_spec/src/open_api/spec.dart:215:72)
#13     fromJsonWithLogging (package:openapi_spec/src/utils/json_logging.dart:4:19)
#14     new OpenApi.fromJson (package:openapi_spec/src/open_api/spec.dart:173:12)
#15     new OpenApi.fromString (package:openapi_spec/src/open_api/spec.dart:164:20)
#16     GenerateCommand.run (package:openapi_spec/src/commands/generate.dart:119:24)
#17     CommandRunner.runCommand (package:args/command_runner.dart:212:27)
#18     CompletionCommandRunner.runCommand (package:cli_completion/src/command_runner/completion_command_runner.dart:89:18)
#19     OpenapiSpecGeneratorCommandRunner.runCommand (package:openapi_spec/src/command_runner.dart:116:30)
#20     OpenapiSpecGeneratorCommandRunner.run (package:openapi_spec/src/command_runner.dart:60:20)
#21     main (file:///C:/Users/dicke/DickerSystems/openapi_spec/bin/openapi_spec.dart:6:66)
#22     _delayEntrypointInvocation.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:295:33)
#23     _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:184:12)

``` 

</p>
</details> 


<details><summary>With Custom Exception</summary>
<p>

Use a custom exception to only log the child fromJson

```
Building package executable... (1.0s)
Built openapi_spec:openapi_spec.
Path to the OpenAPI schema file: .\a.yaml
Unhandled exception:
Failed to parse Tag: {name: 456, description: Learn how to turn audio into text.}  ONLY THE BAD JSON
type 'int' is not a subtype of type 'String' in type cast
#0      fromJsonWithLogging (package:openapi_spec/src/utils/json_logging.dart:10:7)
#1      new Tag.fromJson (package:openapi_spec/src/open_api/tag.dart:26:7)
#2      new OpenApi.fromJson.<anonymous closure>.<anonymous closure> (package:openapi_spec/src/open_api/spec.dart:215:59)
#3      MappedListIterable.elementAt (dart:_internal/iterable.dart:425:31)
#4      ListIterator.moveNext (dart:_internal/iterable.dart:354:26)
#5      new _GrowableList._ofEfficientLengthIterable (dart:core-patch/growable_array.dart:189:27)
#6      new _GrowableList.of (dart:core-patch/growable_array.dart:150:28)
#7      new List.of (dart:core-patch/array_patch.dart:39:18)
#8      ListIterable.toList (dart:_internal/iterable.dart:224:7)
#9      new OpenApi.fromJson.<anonymous closure> (package:openapi_spec/src/open_api/spec.dart:215:72)
#10     fromJsonWithLogging (package:openapi_spec/src/utils/json_logging.dart:4:19)
#11     new OpenApi.fromJson (package:openapi_spec/src/open_api/spec.dart:173:12)
#12     new OpenApi.fromString (package:openapi_spec/src/open_api/spec.dart:164:20)
#13     GenerateCommand.run (package:openapi_spec/src/commands/generate.dart:119:24)
#14     CommandRunner.runCommand (package:args/command_runner.dart:212:27)
#15     CompletionCommandRunner.runCommand (package:cli_completion/src/command_runner/completion_command_runner.dart:89:18)
#16     OpenapiSpecGeneratorCommandRunner.runCommand (package:openapi_spec/src/command_runner.dart:116:30)
#17     OpenapiSpecGeneratorCommandRunner.run (package:openapi_spec/src/command_runner.dart:60:20)
#18     main (file:///C:/Users/dicke/DickerSystems/openapi_spec/bin/openapi_spec.dart:6:66)
#19     _delayEntrypointInvocation.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:295:33)
#20     _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:184:12)
``` 

</p>
</details>

**Review by walsha2 — APPROVED** [review]
LGTM

**Pull request merged**